Scripts/Python/plasma/generate_stubs.py

```diff
@@ -0,0 +1,382 @@
+                # and are assumed to be always defined already.
+                if base.__module__ == value.__module__ and base not in already_seen_classes
+            ]
+
+            if missing_bases:
+                # Remove all of the missing base classes from their original list position...
+                for missing_base_item in missing_bases:
+                    attrs.remove(missing_base_item)
+                # (...verify that we didn't somehow remove anything at or before index i...)
+                assert attrs[i] == (name, value)
+                # ...and re-add them right before the current class.
+                attrs[i:i] = missing_bases
+
+                # We intentionally continue WITHOUT incrementing i here!
+                # The next iteration must process the classes that we just moved around,
+                # so that if the bases themselves have any missing bases,
+                # those are handled recursively.
+                # Iteration continues normally once the class at index i has no more missing bases.
+                continue
+            else:
+                # All base classes are already defined - nothing more to be done.
+                already_seen_classes.add(value)
+
+        i += 1
+
+    return attrs
+
+def parse_type_from_doc(doc: str) -> tuple[str, str]:
+    if doc is None:
+        doc = ""
+
+    if doc.startswith(docstring_type_prefix):
+        # The "Type: " prefix is used for both functions/methods and properties.
+        # For functions, the prefix is followed by a complete function signature,
+        # including parentheses, parameter names, and (optionally) parameter and return types.
+        # Example: "Type: (x: int, y: int) -> int"
+        # For properties, the prefix is followed by a single type.
+        # Example: "Type: Tuple[str, int]"
+        type_line, _, doc_body = doc.partition("\n")
+        type_string = type_line.removeprefix(docstring_type_prefix)
+        return type_string, doc_body
+    elif doc.startswith(docstring_params_prefix):
+        # Cyan's original format for function/method signatures.
+        # The "Params: " prefix is followed by an unparenthesized list of parameter names.
+        # Example: "Params: x,y"
+        params_line, _, doc_body = doc.partition("\n")
+        params = params_line.removeprefix(docstring_params_prefix)
+        return f"({params})", doc_body
+    else:
+        return "", doc
+
+def format_qualified_name(cls: type, context_module_name: str) -> str:
+    if cls.__module__ in {"builtins", context_module_name}:
+        return cls.__qualname__
+    else:
+        return cls.__module__ + "." + cls.__qualname__
+
+def add_indents(indent: str, lines: Iterable[str]) -> Iterable[str]:
+    for line in lines:
+        if line:
+            yield indent + line
+        else:
+            # Don't add extra whitespace on blank lines.
+            yield ""
```

Comment:
This could be replaced with [`textwrap.indent()`](https://docs.python.org/3/library/textwrap.html#module-textwrap).

Reply:
This is a good idea, but I would prefer to do this in a later PR, because it changes the formatting of docstrings that contain newlines. It turns out we have quite a few of those and it would clutter up the diff with even more whitespace changes.

Reply:
Interesting. I thought based on the documentation that it would give the same result, especially if we provided a predicate, eg

```python
add_indents = functools.partial(textwrap.indent, predicate=lambda line: bool(line))
```

(The call to `bool()` is for the sake of clarity and could be omitted.)

Reply:
With that being said, with the need for `functools.partial()`, we might be reaching too far into the bag of tricks. I think I'd be okay with keeping `add_indents()` if we remove the string concat in favor of an f-string 😉

Reply:
The "problem" is that `textwrap.indent` will indent every line, whereas my `add_indents` lets you insert newlines that will not cause an indent, e. g. `add_indents("    ", ["indented\nnot indented", "indented again"])`.

This behavior was by accident, but it turned out to be necessary for matching Cyan's original formatting, where multiline docstrings only have their first line indented and not the following lines.

Reply:
Makes sense. Thanks for the clarification